File: shared/core/src/commonMain/kotlin/com/dcp/core/crypto/Envelope.kt
// Explicit-IV AES-GCM (encryptWithIv/decryptWithIv) is a delicate API because
// caller-managed nonces invite reuse. Our nonces are derived deterministically
// from (deviceId, counter) per spec 4.5 precisely so reuse cannot happen; the
// server additionally rejects a repeated (key_id, nonce) pair.
@file:OptIn(DelicateCryptographyApi::class)

package com.dcp.core.crypto

import dev.whyoleg.cryptography.BinarySize.Companion.bytes
import dev.whyoleg.cryptography.DelicateCryptographyApi
import dev.whyoleg.cryptography.CryptographyProvider
import dev.whyoleg.cryptography.algorithms.AES
import dev.whyoleg.cryptography.algorithms.HKDF
import dev.whyoleg.cryptography.algorithms.SHA256
import dev.whyoleg.cryptography.algorithms.XDH
import dev.whyoleg.cryptography.random.CryptographyRandom
import kotlinx.serialization.json.Json
import kotlinx.serialization.json.JsonElement

/**
 * Kotlin implementation of the DCP encryption envelope.
 *
 * The Python module backend/app/modules/crypto/envelope.py is the reference;
 * this implementation must produce byte-identical output for every vector in
 * conformance/crypto. Spec: specs/encryption-envelope-v0.1.md.
 *
 * AAD contents and nonce derivation are normative — never change them here
 * without a spec change, and never without a matching conformance vector.
 *
 * Primitives come from a [CryptographyProvider]: JCA on the JVM, JCA with
 * BouncyCastle on Android (platform JCA lacks X25519 below API 33), CryptoKit
 * on iOS. Same algorithms everywhere: X25519, HKDF-SHA256, AES-256-GCM.
 */

const val ENVELOPE_VERSION: Int = 1

const val CONTENT_KEY_BYTES: Int = 32
const val NONCE_BYTES: Int = 12
const val MEDIA_CHUNK_BYTES: Int = 4 * 1024 * 1024

class EnvelopeException(message: String, cause: Throwable? = null) : Exception(message, cause)

/** A content key wrapped to one recipient project key (spec 4.3). */
class WrappedKey(
    val projectKeyId: String,
    val contentKeyId: String,
    val ephemeralPublic: ByteArray,
    val nonce: ByteArray,
    val wrappedKey: ByteArray,
)

class EncryptedValue(val ciphertext: ByteArray, val nonce: ByteArray)

class EncryptionEnvelope(provider: CryptographyProvider = CryptographyProvider.Default) {

    private val xdh = provider.get(XDH)
    private val aesGcm = provider.get(AES.GCM)
    private val hkdf = provider.get(HKDF)
    private val sha256 = provider.get(SHA256)

    private val wrapInfo = "dcp/v1/wrap".encodeToByteArray()
    private val opNonceInfo = "dcp/v1/op-nonce".encodeToByteArray()
    private val mediaNonceInfo = "dcp/v1/media-nonce".encodeToByteArray()

    // -----------------------------------------------------------------------
    // Nonce derivation (spec 4.5, 6)
    // -----------------------------------------------------------------------

    /**
     * Deterministic per-operation nonce. Safe because each device owns its own
     * content key: a nonce can only repeat if a device reuses a logical
     * counter, which the sync protocol already forbids.
     */
    suspend fun opNonce(deviceId: String, counter: Long): ByteArray {
        if (counter < 0) throw EnvelopeException("counter must be non-negative")
        val material = opNonceInfo + deviceId.encodeToByteArray() + bigEndian(counter)
        return sha256.hasher().hash(material).copyOf(NONCE_BYTES)
    }

    suspend fun mediaNonce(mediaId: String, chunkIndex: Long): ByteArray {
        if (chunkIndex < 0) throw EnvelopeException("chunk index must be non-negative")
        val material = mediaNonceInfo + mediaId.encodeToByteArray() + bigEndian(chunkIndex)
        return sha256.hasher().hash(material).copyOf(NONCE_BYTES)
    }

    // -----------------------------------------------------------------------
    // Key wrapping (spec 4.4)
    // -----------------------------------------------------------------------

    /**
     * Wrap a content key to one recipient.
     *
     * [ephemeralPrivateKey] and [nonce] are injectable so conformance vectors
     * can pin them; production callers must leave both null.
     */
    suspend fun wrapContentKey(
        contentKey: ByteArray,
        contentKeyId: String,
        recipientPublicKey: ByteArray,
        projectKeyId: String,
        ephemeralPrivateKey: ByteArray? = null,
        nonce: ByteArray? = null,
    ): WrappedKey {
        if (contentKey.size != CONTENT_KEY_BYTES) {
            throw EnvelopeException("content key must be $CONTENT_KEY_BYTES bytes")
        }

        val ephemeralPrivate =
            if (ephemeralPrivateKey != null) {
                xdh.privateKeyDecoder(XDH.Curve.X25519)
                    .decodeFromByteArray(XDH.PrivateKey.Format.RAW, ephemeralPrivateKey)
            } else {
                xdh.keyPairGenerator(XDH.Curve.X25519).generateKey().privateKey
            }
        val ephemeralPublic =
            ephemeralPrivate.getPublicKey().encodeToByteArray(XDH.PublicKey.Format.RAW)

        val recipient = xdh.publicKeyDecoder(XDH.Curve.X25519)
            .decodeFromByteArray(XDH.PublicKey.Format.RAW, recipientPublicKey)
        val shared = ephemeralPrivate.sharedSecretGenerator().generateSharedSecretToByteArray(recipient)
        val wrappingKey = wrappingKey(shared, recipientPublicKey, contentKeyId)

        val wrapNonce = nonce ?: CryptographyRandom.nextBytes(NONCE_BYTES)
        val aad = projectKeyId.encodeToByteArray() + contentKeyId.encodeToByteArray()
        val wrapped = aesKey(wrappingKey).cipher().encryptWithIv(wrapNonce, contentKey, aad)

        return WrappedKey(
            projectKeyId = projectKeyId,
            contentKeyId = contentKeyId,
            ephemeralPublic = ephemeralPublic,
            nonce = wrapNonce,
            wrappedKey = wrapped,
        )
    }

    /**
     * Wrap to every active project key (spec 4.3). A lost private key means
     * permanently unrecoverable data; multi-recipient wrapping is the answer.
     */
    suspend fun wrapToRecipients(
        contentKey: ByteArray,
        contentKeyId: String,
        recipients: Map<String, ByteArray>,
    ): List<WrappedKey> {
        if (recipients.isEmpty()) throw EnvelopeException("at least one recipient key is required")
        return recipients.map { (keyId, public) ->
            wrapContentKey(contentKey, contentKeyId, public, keyId)
        }
    }

    suspend fun unwrapContentKey(wrapped: WrappedKey, recipientPrivateKey: ByteArray): ByteArray {
        val private = xdh.privateKeyDecoder(XDH.Curve.X25519)
            .decodeFromByteArray(XDH.PrivateKey.Format.RAW, recipientPrivateKey)
        val recipientPublic = private.getPublicKey().encodeToByteArray(XDH.PublicKey.Format.RAW)

        val ephemeral = xdh.publicKeyDecoder(XDH.Curve.X25519)
            .decodeFromByteArray(XDH.PublicKey.Format.RAW, wrapped.ephemeralPublic)
        val shared = private.sharedSecretGenerator().generateSharedSecretToByteArray(ephemeral)
        val wrappingKey = wrappingKey(shared, recipientPublic, wrapped.contentKeyId)

        val aad = wrapped.projectKeyId.encodeToByteArray() + wrapped.contentKeyId.encodeToByteArray()
        try {
            return aesKey(wrappingKey).cipher().decryptWithIv(wrapped.nonce, wrapped.wrappedKey, aad)
        } catch (cause: Exception) {
            throw EnvelopeException("content key unwrap failed", cause)
        }
    }

    private suspend fun wrappingKey(
        shared: ByteArray,
        recipientPublic: ByteArray,
        contentKeyId: String,
    ): ByteArray =
        hkdf.secretDerivation(
            SHA256,
            CONTENT_KEY_BYTES.bytes,
            recipientPublic,
            wrapInfo + contentKeyId.encodeToByteArray(),
        ).deriveSecretToByteArray(shared)

    // -----------------------------------------------------------------------
    // Operation values (spec 5)
    // -----------------------------------------------------------------------

    /**
     * AAD binds a ciphertext to its exact location. Without `path` a server
     * operator could move an encrypted answer between fields; without
     * `formVersion` a ciphertext could be replayed against a form version
     * where the same path means something else.
     */
    private fun opAad(opId: String, submissionId: String, path: String, formVersion: Int): ByteArray =
        listOf(opId, submissionId, path, formVersion.toString())
            .joinToString("|")
            .encodeToByteArray()

    suspend fun encryptOpValue(
        value: JsonElement,
        contentKey: ByteArray,
        opId: String,
        submissionId: String,
        path: String,
        formVersion: Int,
        deviceId: String,
        counter: Long,
    ): EncryptedValue {
        val nonce = opNonce(deviceId, counter)
        val aad = opAad(opId, submissionId, path, formVersion)
        val ciphertext = aesKey(contentKey).cipher()
            .encryptWithIv(nonce, CanonicalJson.encode(value), aad)
        return EncryptedValue(ciphertext, nonce)
    }

    suspend fun decryptOpValue(
        ciphertext: ByteArray,
        nonce: ByteArray,
        contentKey: ByteArray,
        opId: String,
        submissionId: String,
        path: String,
        formVersion: Int,
    ): JsonElement {
        val aad = opAad(opId, submissionId, path, formVersion)
        val plaintext = try {
            aesKey(contentKey).cipher().decryptWithIv(nonce, ciphertext, aad)
        } catch (cause: Exception) {
            throw EnvelopeException("operation value authentication failed", cause)
        }
        return Json.parseToJsonElement(plaintext.decodeToString())
    }

    // -----------------------------------------------------------------------
    // Media (spec 6)
    // -----------------------------------------------------------------------

    suspend fun encryptMediaChunk(
        chunk: ByteArray,
        mediaKey: ByteArray,
        mediaId: String,
        chunkIndex: Long,
    ): EncryptedValue {
        val nonce = mediaNonce(mediaId, chunkIndex)
        val aad = mediaId.encodeToByteArray() + bigEndian(chunkIndex)
        return EncryptedValue(aesKey(mediaKey).cipher().encryptWithIv(nonce, chunk, aad), nonce)
    }

    suspend fun decryptMediaChunk(
        ciphertext: ByteArray,
        mediaKey: ByteArray,
        mediaId: String,
        chunkIndex: Long,
    ): ByteArray {
        val nonce = mediaNonce(mediaId, chunkIndex)
        val aad = mediaId.encodeToByteArray() + bigEndian(chunkIndex)
        try {
            return aesKey(mediaKey).cipher().decryptWithIv(nonce, ciphertext, aad)
        } catch (cause: Exception) {
            throw EnvelopeException("media chunk authentication failed", cause)
        }
    }

    /**
     * Content address computed over CIPHERTEXT, never plaintext. Hashing
     * plaintext would let the server confirm that two submissions contain the
     * same photograph — exactly the inference this mode exists to prevent.
     */
    fun ciphertextHash(ciphertextChunks: List<ByteArray>): String {
        val hash = sha256.hasher().createHashFunction().use { fn ->
            ciphertextChunks.forEach { fn.update(it) }
            fn.hashToByteArray()
        }
        return hash.joinToString("") { byte -> byte.toUByte().toString(16).padStart(2, '0') }
    }

    // -----------------------------------------------------------------------

    private suspend fun aesKey(raw: ByteArray): AES.GCM.Key =
        aesGcm.keyDecoder().decodeFromByteArray(AES.Key.Format.RAW, raw)

    private fun bigEndian(value: Long): ByteArray =
        ByteArray(8) { i -> (value ushr (8 * (7 - i)) and 0xFF).toByte() }
}

File: backend/app/modules/crypto/envelope.py
"""Reference implementation of the DCP encryption envelope.

Normative for the Kotlin implementation in shared/core. Spec:
specs/encryption-envelope-v0.1.md.

Design summary: the unit of encryption is the operation VALUE, not the
submission, so operation-based sync (resumability, merge, tombstones) keeps
working. Each device holds its own content key per submission, which makes
peer-to-peer contribution possible and lets nonces be derived deterministically
without cross-device coordination.
"""

from __future__ import annotations

import hashlib
import json
from dataclasses import dataclass
from typing import Any

from cryptography.hazmat.primitives import hashes
from cryptography.hazmat.primitives.asymmetric.x25519 import (
    X25519PrivateKey,
    X25519PublicKey,
)
from cryptography.hazmat.primitives.ciphers.aead import AESGCM
from cryptography.hazmat.primitives.kdf.hkdf import HKDF

ENVELOPE_VERSION = 1

WRAP_INFO = b"dcp/v1/wrap"
OP_NONCE_INFO = b"dcp/v1/op-nonce"
MEDIA_NONCE_INFO = b"dcp/v1/media-nonce"

CONTENT_KEY_BYTES = 32
NONCE_BYTES = 12
MEDIA_CHUNK_BYTES = 4 * 1024 * 1024


class EnvelopeError(Exception):
    """Raised when an envelope is malformed, or authentication fails."""


# ---------------------------------------------------------------------------
# Canonical JSON (spec 5.1)
# ---------------------------------------------------------------------------


def canonical_json(value: Any) -> bytes:
    """Serialise a value identically on every platform.

    Without this, the same answer produces different ciphertext on Android and
    iOS and cross-platform test vectors become impossible to write.
    """
    return json.dumps(
        value,
        sort_keys=True,
        separators=(",", ":"),
        ensure_ascii=False,
        allow_nan=False,
    ).encode("utf-8")


# ---------------------------------------------------------------------------
# Nonce derivation (spec 4.5, 6)
# ---------------------------------------------------------------------------


def op_nonce(device_id: str, counter: int) -> bytes:
    """Deterministic per-operation nonce.

    Safe because each device owns its own content key: a nonce can only repeat
    if a device reuses a logical counter, which the sync protocol already
    forbids. Derivation avoids depending on device RNG quality, which on cheap
    Android hardware is not something to rely on.
    """
    if counter < 0:
        raise EnvelopeError("counter must be non-negative")
    material = OP_NONCE_INFO + device_id.encode("utf-8") + counter.to_bytes(8, "big")
    return hashlib.sha256(material).digest()[:NONCE_BYTES]


def media_nonce(media_id: str, chunk_index: int) -> bytes:
    material = (
        MEDIA_NONCE_INFO + media_id.encode("utf-8") + chunk_index.to_bytes(8, "big")
    )
    return hashlib.sha256(material).digest()[:NONCE_BYTES]


# ---------------------------------------------------------------------------
# Key wrapping (spec 4.4)
# ---------------------------------------------------------------------------


@dataclass(frozen=True)
class WrappedKey:
    project_key_id: str
    content_key_id: str
    ephemeral_public: bytes
    nonce: bytes
    wrapped_key: bytes

    def to_dict(self) -> dict[str, Any]:
        return {
            "projectKeyId": self.project_key_id,
            "contentKeyId": self.content_key_id,
            "ephemeralPublic": self.ephemeral_public.hex(),
            "nonce": self.nonce.hex(),
            "wrappedKey": self.wrapped_key.hex(),
        }


def _wrapping_key(shared: bytes, recipient_public: bytes, content_key_id: str) -> bytes:
    return HKDF(
        algorithm=hashes.SHA256(),
        length=32,
        salt=recipient_public,
        info=WRAP_INFO + content_key_id.encode("utf-8"),
    ).derive(shared)


def wrap_content_key(
    content_key: bytes,
    content_key_id: str,
    recipient_public_key: bytes,
    project_key_id: str,
    *,
    ephemeral_private: X25519PrivateKey | None = None,
    nonce: bytes | None = None,
) -> WrappedKey:
    """Wrap a content key to one recipient.

    ``ephemeral_private`` and ``nonce`` are injectable so conformance vectors
    can pin them; production callers must leave both as None.
    """
    if len(content_key) != CONTENT_KEY_BYTES:
        raise EnvelopeError(f"content key must be {CONTENT_KEY_BYTES} bytes")

    ephemeral_private = ephemeral_private or X25519PrivateKey.generate()
    ephemeral_public = ephemeral_private.public_key().public_bytes_raw()

    shared = ephemeral_private.exchange(X25519PublicKey.from_public_bytes(recipient_public_key))
    wrapping_key = _wrapping_key(shared, recipient_public_key, content_key_id)

    nonce = nonce or AESGCM.generate_key(bit_length=128)[:NONCE_BYTES]
    aad = project_key_id.encode("utf-8") + content_key_id.encode("utf-8")
    wrapped = AESGCM(wrapping_key).encrypt(nonce, content_key, aad)

    return WrappedKey(
        project_key_id=project_key_id,
        content_key_id=content_key_id,
        ephemeral_public=ephemeral_public,
        nonce=nonce,
        wrapped_key=wrapped,
    )


def wrap_to_recipients(
    content_key: bytes,
    content_key_id: str,
    recipients: dict[str, bytes],
) -> list[WrappedKey]:
    """Wrap to every active project key (spec 4.3).

    Lost private key means permanently unrecoverable data. Multi-recipient
    wrapping is the answer to that; a server-side escrow would reintroduce the
    trust the mode exists to remove.
    """
    if not recipients:
        raise EnvelopeError("at least one recipient key is required")
    return [
        wrap_content_key(content_key, content_key_id, public, key_id)
        for key_id, public in recipients.items()
    ]


def unwrap_content_key(wrapped: WrappedKey, recipient_private_key: bytes) -> bytes:
    private = X25519PrivateKey.from_private_bytes(recipient_private_key)
    recipient_public = private.public_key().public_bytes_raw()

    shared = private.exchange(X25519PublicKey.from_public_bytes(wrapped.ephemeral_public))
    wrapping_key = _wrapping_key(shared, recipient_public, wrapped.content_key_id)

    aad = wrapped.project_key_id.encode("utf-8") + wrapped.content_key_id.encode("utf-8")
    try:
        return AESGCM(wrapping_key).decrypt(wrapped.nonce, wrapped.wrapped_key, aad)
    except Exception as exc:
        raise EnvelopeError("content key unwrap failed") from exc


# ---------------------------------------------------------------------------
# Operation values (spec 5)
# ---------------------------------------------------------------------------


def _op_aad(op_id: str, submission_id: str, path: str, form_version: int) -> bytes:
    """Bind a ciphertext to its exact location.

    Without `path` in the AAD, a server operator could move an encrypted answer
    from `income` to `age` and the client would decrypt it without complaint.
    Without `form_version`, a ciphertext could be replayed against a version of
    the form where the same path means something else.
    """
    return b"|".join(
        [
            op_id.encode("utf-8"),
            submission_id.encode("utf-8"),
            path.encode("utf-8"),
            str(form_version).encode("utf-8"),
        ]
    )


def encrypt_op_value(
    value: Any,
    content_key: bytes,
    *,
    op_id: str,
    submission_id: str,
    path: str,
    form_version: int,
    device_id: str,
    counter: int,
) -> tuple[bytes, bytes]:
    """Encrypt one operation value. Returns (ciphertext, nonce)."""
    nonce = op_nonce(device_id, counter)
    aad = _op_aad(op_id, submission_id, path, form_version)
    ciphertext = AESGCM(content_key).encrypt(nonce, canonical_json(value), aad)
    return ciphertext, nonce


def decrypt_op_value(
    ciphertext: bytes,
    nonce: bytes,
    content_key: bytes,
    *,
    op_id: str,
    submission_id: str,
    path: str,
    form_version: int,
) -> Any:
    aad = _op_aad(op_id, submission_id, path, form_version)
    try:
        plaintext = AESGCM(content_key).decrypt(nonce, ciphertext, aad)
    except Exception as exc:
        raise EnvelopeError("operation value authentication failed") from exc
    return json.loads(plaintext.decode("utf-8"))


# ---------------------------------------------------------------------------
# Media (spec 6)
# ---------------------------------------------------------------------------


def encrypt_media_chunk(
    chunk: bytes, media_key: bytes, *, media_id: str, chunk_index: int
) -> tuple[bytes, bytes]:
    nonce = media_nonce(media_id, chunk_index)
    aad = media_id.encode("utf-8") + chunk_index.to_bytes(8, "big")
    return AESGCM(media_key).encrypt(nonce, chunk, aad), nonce


def decrypt_media_chunk(
    ciphertext: bytes, media_key: bytes, *, media_id: str, chunk_index: int
) -> bytes:
    nonce = media_nonce(media_id, chunk_index)
    aad = media_id.encode("utf-8") + chunk_index.to_bytes(8, "big")
    try:
        return AESGCM(media_key).decrypt(nonce, ciphertext, aad)
    except Exception as exc:
        raise EnvelopeError("media chunk authentication failed") from exc


def ciphertext_hash(ciphertext_chunks: list[bytes]) -> str:
    """Content address computed over CIPHERTEXT, never plaintext.

    Hashing plaintext would let the server confirm that two submissions contain
    the same photograph — exactly the inference end-to-end encryption exists to
    prevent.
    """
    digest = hashlib.sha256()
    for chunk in ciphertext_chunks:
        digest.update(chunk)
    return digest.hexdigest()


# ---------------------------------------------------------------------------
# Sensitivity propagation (spec 5.2)
# ---------------------------------------------------------------------------


def check_sensitivity_propagation(compiled_form: Any) -> list[str]:
    """A calculated field reading a sensitive field must itself be sensitive.

    Otherwise the calculation leaks its input. Returns violation messages; an
    empty list means the form is safe to publish in field_level mode.
    """
    violations: list[str] = []
    for field_id, field in compiled_form.fields.items():
        if field.node.get("sensitive") is True:
            continue
        for dep in field.depends_on:
            base = dep.split("[")[0].split(".")[-1] if "]." in dep else dep
            dep_field = compiled_form.fields.get(base)
            if dep_field is not None and dep_field.node.get("sensitive") is True:
                violations.append(
                    f"{field_id!r} is not sensitive but depends on sensitive field {base!r}"
                )
    return violations
